flow/controllers/imitation_learning/imitating_network.py

```diff
@@ -0,0 +1,267 @@
+
+        self.replay_buffer = ReplayBuffer(replay_buffer_size)
+
+    def build_network(self):
+        """
+        Defines neural network for choosing actions. Defines placeholders and forward pass
+        """
+        # setup placeholders for network input and labels for training, and hidden layers/output
+        if self.stochastic:
+            self.model = build_neural_net_stochastic(self.obs_dim, self.action_dim, self.fcnet_hiddens)
+        else:
+            self.model = build_neural_net_deterministic(self.obs_dim, self.action_dim, self.fcnet_hiddens)
+
+
+    def compile_network(self):
+        """
+        Compiles Keras network with appropriate loss and optimizer
+        """
+        loss = get_loss(self.stochastic, self.variance_regularizer)
+        self.model.compile(loss=loss, optimizer='adam')
+
+
+    def train(self, observation_batch, action_batch):
+        """
+        Executes one training (gradient) step for the given batch of observation and action data
+
+        Parameters
+        ----------
+        observation_batch : numpy array
+            numpy array containing batch of observations (inputs)
+        action_batch : numpy array
+            numpy array containing batch of actions (labels)
+        """
+
+        # reshape action_batch to ensure a shape (batch_size, action_dim)
+        action_batch = action_batch.reshape(action_batch.shape[0], self.action_dim)
+        # one gradient step on batch
+        loss = self.model.train_on_batch(observation_batch, action_batch)
+
+        # tensorboard
+        summary = tf.Summary(value=[tf.Summary.Value(tag="imitation training loss", simple_value=loss), ])
+        self.writer.add_summary(summary, global_step=self.train_steps)
+        self.train_steps += 1
+
+    def get_accel_from_observation(self, observation):
+        """
+        Gets the network's acceleration prediction based on given observation/state
+
+        Parameters
+        ----------
+        observation : numpy array
+            numpy array containing a single observation
+
+        Returns
+        -------
+        numpy array
+            one element numpy array containing acceleration
+        """
+
+        # network expects an array of arrays (matrix); if single observation (no batch), convert to array of arrays
+        if len(observation.shape)<=1:
+            observation = observation[None]
+        # "batch size" is 1, so just get single acceleration/acceleration vector
+        network_output = self.model.predict(observation)
+        if self.stochastic:
+            mean, log_std = network_output[:, :self.action_dim], network_output[:, self.action_dim:]
+            var = np.exp(2 * log_std)
+
+            variance_norm = np.linalg.norm(var)
+            summary = tf.Summary(value=[tf.Summary.Value(tag="Variance norm", simple_value=variance_norm), ])
+            self.writer.add_summary(summary, global_step=self.action_steps)
+
+            cov_matrix = np.diag(var[0])
```

Comment:
why var[0]?

Reply:
Magic number means you should write what the expected dimension of this object is. This makes it so at least the magic number is understandable